Enable semi-join optimization for left-deep trees through bloom filters

saketj opened the pull request:
This PR adds support for semi-join optimization for left-deep trees through the use of bloom filters. The motivation behind this PR is that often the bottom-most join in a left-deep tree returns maximum number of tuples from the fact table, a majority of which are then discarded by joins up the tree. If somehow, the selectivity information from up the tree could be passed down to the bottom-most operators, we could avoid the cost of potentially passing and storing huge number of unnecessary tuples during query execution. 

Bloom filters are lightweight data structures that can help us push down this selectivity information down the operator tree. Therefore, the key idea is to delay the processing of the bottom-most probe operator on the fact table, until bloom filters on all the build operators on all the dimension tables are built. Then pass the bloom filters sideways to this probe operator, which can use these bloom filters to weed out the unnecessary tuples.

This feature is controlled by GFLAGs and is turned off by default. Pass the `--optimize_joins` flag from the command-line to explicitly enable this optimization.

On a typical left-deep join query like SSB Query 4, this optimization reduces the query execution time by almost 40%.

hbdeshmukh commented on query_execution/QueryContext.hpp:
```diff
@@ -181,6 +189,51 @@ class QueryContext {
+   *
+   * @return The BloomFilter, already created in the constructor.
+   **/
+  inline const BloomFilter* getBloomFilter(const bloom_filter_id id) const {
```
This function returns a pointer but the doxygen says const reference.

saketj replied:
Updated the comment.

hbdeshmukh commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -563,12 +572,33 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
       physical_plan->left_join_attributes();
   for (const E::AttributeReferencePtr &left_join_attribute : left_join_attributes) {
+    // Try to determine the original relation referenced in the Hash Join.
+    const CatalogRelation *catalogRelation =
```
Variable names are usually ``catalog_relation`` instead of ``catalogRelation``. Also, can you think of a more informative name for this variable?

saketj replied:
Fixed.

hbdeshmukh commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,104 @@
+
+#include <cstddef>
+#include <vector>
+#include <utility>
```
Please sort the included files.

saketj replied:
Done.

hbdeshmukh commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,104 @@
+
+  // The optimization is done by modifying each of the build operators in the chain
+  // to generate a bloom filter on the build key during their hash table creation.
+  // The origin(bottom-most) probe operator is then modified to query all the bloom
```
You can say leaf-level probe operator.

saketj replied:
Updated.

hbdeshmukh commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,104 @@
+  // The optimization is done by modifying each of the build operators in the chain
+  // to generate a bloom filter on the build key during their hash table creation.
+  // The origin(bottom-most) probe operator is then modified to query all the bloom
+  // filters generated from all the build operators in the chain. Bloom filter queries
```
What do you mean by bloom filter query?

saketj replied:
It meant to query the bloom filter for membership test. I have updated the comments to be more precise here. Thanks.

hbdeshmukh commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+  HashJoinInfo(QueryPlan::DAGNodeIndex build_operator_index,
+               QueryPlan::DAGNodeIndex join_operator_index,
+               const relation_id &build_relation_id,
+               const relation_id &probe_relation_id,
```
I am wondering if it's worth using a reference for an integer vs using the integer itself. If int takes up four bytes and a reference is 8 bytes, then we should go for the shorter one, I feel.

saketj replied:
Removed the reference usage.

hbdeshmukh commented on relational_operators/CMakeLists.txt:
```diff
@@ -92,7 +92,7 @@ target_link_libraries(quickstep_relationaloperators_BuildHashOperator
                       quickstep_storage_TupleReference
                       quickstep_storage_TupleStorageSubBlock
-                      quickstep_storage_ValueAccessor
+                      quickstep_storage_ValueAccessor                      
```
Trailing spaces.

saketj replied:
Fixed.

hbdeshmukh commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,104 @@
+  // are issued just prior to probing the hash table.
+
+  QueryPlan::DAGNodeIndex origin_node = 0;
+  while (origin_node < hash_joins_.size() - 1) {
```
This function is somewhat complicated to understand. I will find you this afternoon, if you are in the office and we can discuss this in person.

saketj commented:
@hbdeshmukh  Thanks Harshad for the comments. I have addressed them now. Sure, I will be there around in the afternoon at the department. We can discuss it the optimization function.

hbdeshmukh commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+   **/
+  inline void addHashJoinInfo(QueryPlan::DAGNodeIndex build_operator_index,
+                              QueryPlan::DAGNodeIndex join_operator_index,
+                              const relation_id build_relation_id,
```
Please rename this variable for better clarity.

hbdeshmukh commented:
I and @saketj had a discussion. Saket will improve the documentation and variable names for better clarity. We also had a discussion on how to write tests for this code.

pateljm commented:
Sounds good @hbdeshmukh. Thanks! 

Thanks @saketj -- This is an awesome feature!

hbdeshmukh commented on storage/HashTable.hpp:
```diff
@@ -1429,6 +1492,11 @@ HashTablePutResult HashTable<ValueT, resizable, serializable, force_key_copy, al
                                        using_prealloc ? &prealloc_state : nullptr);
+            // Insert into bloom filter, if enabled.
+            if (has_build_side_bloom_filter_) {
+              this->build_bloom_filter_->insert(static_cast<const std::uint8_t *>(key.getDataPtr()),
```
I am not completely sure if this code will be thread safe. I think we should have another round of discussion to discuss the thread-safety of using bloom filters inside the hash table.

saketj replied:
@hbdeshmukh Harshad, thread safety is being guaranteed here from inside the `build_bloom_filter_->insert()`, which implements an exclusive mutex lock to allow only one thread to modify the bloom filter.
For containment test, I guess we will not require a lock, since the assumption is that a bloom filter will never be used, unless it has been built.

hbdeshmukh replied:
I see, thanks for that information. However, we are inserting an entry in the bloom filter for every tuple, and while doing so, acquiring a lock on the bloom filter. Even though the time to insert an entry in the bloom filter may not be high (flipping a single bit), we are essentially serializing all the worker threads building the hash table. 

I feel that there are some relatively easy optimizations possible, and it may be worth discussing them.

saketj replied:
Thanks. That's a good point. I guess inserting into a local bloom filter and doing a union with the shared bloom filter with locking may solve this problem. Sure, I will stop by today in the afternoon to discuss this in detail.

zuyu commented on query_execution/QueryContext.cpp:
```diff
@@ -65,6 +68,13 @@ QueryContext::QueryContext(const serialization::QueryContext &proto,
 
+  for (int i = 0; i < proto.bloom_filters_size(); ++i) {
+    const serialization::BloomFilter &bloom_filter_proto = proto.bloom_filters(i);
+    bloom_filters_.emplace_back(new BloomFilter(bloom_filter_proto.bloom_filter_seed(),
```
Could we have a `BloomFilter` constructor that takes its proto as the single argument? Thus, this line could be simplified to

`bloom_filters_.emplace_back(new BloomFilter(bloom_filter_proto));`

saketj replied:
Added a new constructor to initialize from the bloom filter protobuf.

zuyu commented on query_execution/QueryContext.cpp:
```diff
@@ -77,6 +87,33 @@ QueryContext::QueryContext(const serialization::QueryContext &proto,
+          probe_attribute_ids.push_back(probe_attribute_id);
+        }
+        join_hash_tables_[i]->addProbeSideAttributeIds(std::move(probe_attribute_ids));
+      }
```
Have you considered of moving some of the code here inside `JoinHashTableFactory ::CreateResizableFromProto`? For example, calling `setBuildSideBloomFilter`.

saketj replied:
@zuyu These functions (e.g. `addProdeSideBloomFilter(), setBuildSideBloomFilter()`) may not always be called and therefore, I thought we can save the cost of passing references to the data structures of Query Context here, when invoking `JoinHashTableFactory ::CreateResizableFromProto()`

zuyu replied:
Sorry, I did not get that.

The goal is to make `QueryContext` clean with a few reconstruction methods.

All the methods called here are setting up for `join_hash_tables_[i]`, reconstructed by `JoinHashTableFactory::CreateResizableFromProto`, based on `proto.join_hash_tables(i)`. So, I was wondering what additional info we need from `QueryContext`.

saketj replied:
@zuyu The information about bloom filters is some additional info that we need to pass from the `QueryContext`. But, nevermind, I liked your idea of keeping the `QueryContext` cleaner and I have refactored the code now to move all these initializations inside `JoinHashTableFactory::CreateResizableFromProto`.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -563,12 +572,33 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
+    } else {
+      probe_relation_id = referenced_stored_catalog_relation->getID();
+      const attribute_id probe_operator_attribute_id =
+      referenced_stored_catalog_relation->getAttributeByName(left_join_attribute->attribute_name())->getID();
```
According to the code style, please add four whitespace indentations.

https://github.com/pivotalsoftware/quickstep/wiki/Quickstep-Style-Guide

saketj replied:
Fixed.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -581,6 +611,19 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
+    } else {
+      build_relation_id = referenced_stored_catalog_relation->getID();
+      const attribute_id build_operator_attribute_id =
+      referenced_stored_catalog_relation->getAttributeByName(right_join_attribute->attribute_name())->getID();
```
Ditto for breaking a line with the four whitespace indentation.

saketj replied:
Fixed.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+class HashJoinInfo {
+ public:
+  HashJoinInfo(QueryPlan::DAGNodeIndex build_operator_index,
+               QueryPlan::DAGNodeIndex join_operator_index,
```
We could also mark two `operator_index`s above `const`.

saketj replied:
Updated.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+               const relation_id &probe_relation_id,
+               std::vector<attribute_id> &&build_attributes,
+               std::vector<attribute_id> &&probe_attributes,
+               const QueryContext::join_hash_table_id &join_hash_table_id)
```
I agree, it does not worth to use a reference for an argument in a plain type.

So, please remove the reference here as well.

saketj replied:
Removed.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+      build_relation_id_(build_relation_id),
+      probe_relation_id_(probe_relation_id),
+      build_attributes_(build_attributes),
+      probe_attributes_(probe_attributes),
```
Please use the correct way for `rvalue`:

`probe_attributes_(std::move(probe_attributes)),`

saketj replied:
Thanks. Updated these.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+#include "glog/logging.h"
+
+namespace quickstep {
+
```
Please remove this empty line.

saketj replied:
Removed.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,137 @@
+                              const relation_id build_relation_id,
+                              const relation_id probe_relation_id,
+                              std::vector<attribute_id> *build_attributes,
+                              std::vector<attribute_id> *probe_attributes,
```
Please use `rvalue` properly:

`std::vector<attribute_id> &&build_attributes,`

And in the caller site:

`std::move(build_attributes),`

saketj replied:
Thanks. Updated.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    // Update the origin node.
+    origin_node = chained_nodes.at(chained_nodes.size() - 1) + 1;
```
Could we refactor to the following?

`origin_node = chained_nodes.back() + 1;`

saketj replied:
Refactored.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,105 @@
+    // Only chains of length greater than one are suitable candidates for semi-join optimization.
+    if (chained_nodes.size() > 1) {
+      std::vector<std::pair<QueryContext::bloom_filter_id, std::vector<attribute_id>>> probe_bloom_filter_info;
+      for (std::size_t &i : chained_nodes) {
```
How about the following refactor:

`for (const std::size_t node : chained_nodes) {`

saketj replied:
Refactored.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,105 @@
+
+    // Only chains of length greater than one are suitable candidates for semi-join optimization.
+    if (chained_nodes.size() > 1) {
+      std::vector<std::pair<QueryContext::bloom_filter_id, std::vector<attribute_id>>> probe_bloom_filter_info;
```
What if we change to `std::unordered_map<QueryContext::bloom_filter_id, std::vector<attribute_id>>`?

saketj replied:
Changed.

zuyu commented on storage/HashTable.hpp:
```diff
@@ -990,6 +991,61 @@ class HashTable : public HashTableBase<resizable,
+   * @param probe_attribute_ids The vector of attribute ids to use for probing
+   *        the bloom filter.
+   **/
+  inline void addProbeSideAttributeIds(std::vector<attribute_id> &&probe_attribute_ids) {
```
I don't understand why using `rvalue` here? Actually, I was thinking if we could initialize these values in the constructor or the factory method.

saketj replied:
@zuyu  These may not be always initialized and hence, I thought to add them as separate functions.

zuyu replied:
Sorry, I did not get that. If we do not have probe side attribute ids, we could leave the `vector` as an empty one.

If I understand correctly, this entry is set once, and won't change. Then we could even mark it as `const` during the reconstruction using `JoinHashTableFactory::CreateResizableFromProto`.

saketj replied:
Moved it now inside `JoinHashTableFactory::CreateResizableFromProto`

zuyu commented on utility/BloomFilter.hpp:
```diff
@@ -47,7 +53,30 @@ class BloomFilter {
+        array_size_in_bytes_(bit_array_size_in_bytes),
+        is_bit_array_owner_(true) {
+    array_size_ = bit_array_size_in_bytes * kNumBitsPerByte;
+    bit_array_.reset(new std::uint8_t[bit_array_size_in_bytes]);
```
We could move both `array_size_` and `bit_array_` in the init list of the constructor as well.

saketj replied:
Moved.

zuyu commented on utility/BloomFilter.hpp:
```diff
@@ -47,7 +53,30 @@ class BloomFilter {
+        is_bit_array_owner_(true) {
+    array_size_ = bit_array_size_in_bytes * kNumBitsPerByte;
+    bit_array_.reset(new std::uint8_t[bit_array_size_in_bytes]);
+    reset();
```
Do we call `reset` more than once? If no, as the implementation of `reset` is simple, so we could move the code in the constructor and get rid of this method.

saketj replied:
@zuyu The reset() function is also used by the BloomFilterStorageSubBlock to clear out the bloom filter, hence added as a separate method.

zuyu commented on utility/BloomFilter.proto:
```diff
@@ -0,0 +1,30 @@
+//   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+//    University of Wisconsin—Madison.
```
For consistency, please add one more whitespace indentation.

saketj replied:
Updated.

zuyu commented on utility/BloomFilter.proto:
```diff
@@ -0,0 +1,30 @@
+  // - Default bloom filter size = 10 KB.
+  // - Default number of hash functions used in bloom filter = 5.
+  required fixed64 bloom_filter_seed = 1 [default = 0xA5A5A5A55A5A5A5A];
+  required fixed64 bloom_filter_size = 2 [default = 10000];
```
I was wondering why we need a `fixed64` here?

saketj replied:
Changed to `uint32`. Thanks

zuyu commented on utility/BloomFilter.proto:
```diff
@@ -0,0 +1,30 @@
+  // - Default number of hash functions used in bloom filter = 5.
+  required fixed64 bloom_filter_seed = 1 [default = 0xA5A5A5A55A5A5A5A];
+  required fixed64 bloom_filter_size = 2 [default = 10000];
+  required int32 number_of_hashes = 3 [default = 5];
```
Could we use `uint32` instead?

zuyu commented:
@saketj I was wondering, as an alternative to add a new `execution_heuristics_`, we just treat this as a new method called `optimizeHashJoin` in the existing `ExecutionGeneerator`.

saketj commented:
@hbdeshmukh and I had discussion on how we can reduce lock contention between various worker threads during bloom filter insertion. An improved locking mechanism is now implemented in the latest commit.

saketj commented:
@zuyu Thanks Zuyu for the comments. I am working on incorporating them into this PR. 
As for ExecutionHeuristics class, I can foresee it as something that will grow to incorporate many more complex optimization logic, along with its own data structures. For example, the logic for identification of left-deep trees is currently simple and ignores some complicated scenarios. This will be work of subsequent PRs, I guess. Also, setting bloom filter properties based on the relation statistics is  another component that has been introduced.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -101,5 +104,23 @@ void ExecutionHeuristics::optimizeExecutionPlan(QueryPlan *query_plan,
+  if (cardinality < kOneThousand) {
+    bloom_filter_proto->set_bloom_filter_size(kOneThousand / kCompressionFactor);
+    bloom_filter_proto->set_number_of_hashes(kVeryLowSparsityHash);
+  } else if (cardinality >= kOneThousand && cardinality < kTenThousand) {
```
We could simplify to ` else if (cardinality < kTenThousand)`, as `cardinality >= kOneThousand` is always `true`.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -101,5 +104,23 @@ void ExecutionHeuristics::optimizeExecutionPlan(QueryPlan *query_plan,
+  } else if (cardinality >= kTenThousand && cardinality < kHundredThousand) {
+    bloom_filter_proto->set_bloom_filter_size(kHundredThousand / kCompressionFactor);
+    bloom_filter_proto->set_number_of_hashes(kMediumSparsityHash);
+  } else if (cardinality >= kHundredThousand) {
```
How about using `else`, instead of `else if`?

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -101,5 +104,23 @@ void ExecutionHeuristics::optimizeExecutionPlan(QueryPlan *query_plan,
+  } else if (cardinality >= kOneThousand && cardinality < kTenThousand) {
+    bloom_filter_proto->set_bloom_filter_size(kTenThousand / kCompressionFactor);
+    bloom_filter_proto->set_number_of_hashes(kLowSparsityHash);
+  } else if (cardinality >= kTenThousand && cardinality < kHundredThousand) {
```
We could remove the always-true `cardinality >= kTenThousand` expression here.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,161 @@
+  static const std::size_t kVeryLowSparsityHash = 1;
+  static const std::size_t kLowSparsityHash = 2;
+  static const std::size_t kMediumSparsityHash = 5;
+  static const std::size_t kHighSparsityHash = 10;
```
I was wondering if we could set above five constants from `kCompressionFactor` to `kHighSparsityHash` as `gflag` variables with the same default values as you set, so that we could benchmark various combinations without recompiling the code.

saketj replied:
Thanks. This is a good suggestion. I am still in process of figuring out which ones of these will be useful in benchmarking. Will update them to GFLAGS as necessary.

zuyu commented on query_optimizer/ExecutionHeuristics.cpp:
```diff
@@ -0,0 +1,126 @@
+    bloom_filter_proto->set_bloom_filter_size(kHundredThousand / kCompressionFactor);
+    bloom_filter_proto->set_number_of_hashes(kMediumSparsityHash);
+  } else if (cardinality >= kHundredThousand) {
+    bloom_filter_proto->set_bloom_filter_size(kMillion / kCompressionFactor);
```
Also, you may want to pre-compute the bloom filter size for each case, to avoid do so for each bloom filter, although it is still lightweight.

zuyu commented on query_optimizer/ExecutionHeuristics.hpp:
```diff
@@ -0,0 +1,161 @@
+ * @brief A simple internal class that holds information about various
+ *        hash joins within the execution plan for a query.
+ **/
+class HashJoinInfo {
```
How about moving this class inside `ExecutionHeuristics`? Also, I notice that you did not mark all data members `private`. If you intend to do so, how about using `struct` instead?

saketj replied:
Moved it inside the `ExecutionHeuristics` class and changed to `struct` now.

saketj commented:
@zuyu @hbdeshmukh  I have addressed all of the comments now. I have also added the unit tests for this PR at `query_optimizer/tests/ExecutionHeuristics_unittest.cpp`. Please let me know if this all looks good and if we can merge.
There is one more incremental feature that builds on top of this PR and it adds the ability to push the bloom filters even further down the tree to a selection operator preceding a probe. I have the changes in my local copy and I am planning to immediately open that PR, once this gets merged. Thanks.

zuyu commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,307 @@
+    std::vector<attribute_id> build_attribute_ids;
+    build_attribute_ids.push_back(build_attribute_id);
+    std::vector<attribute_id> probe_attribute_ids;
+    probe_attribute_ids.push_back(probe_attribute_id);
```
We could refactor to

```
  std::vector<attribute_id> build_attribute_ids(1, build_attribute_id);
  std::vector<attribute_id> probe_attribute_ids(1, probe_attribute_id);
```

saketj replied:
Refactored.

zuyu commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,307 @@
+                                                       const attribute_id build_attribute_id,
+                                                       const QueryContext::join_hash_table_id join_hash_table_index) {
+    std::vector<attribute_id> build_attribute_ids;
+    build_attribute_ids.push_back(build_attribute_id);
```
We could refactor to `const std::vector<attribute_id> build_attribute_ids(1, build_attribute_id);`

saketj replied:
Refactored.

zuyu commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,307 @@
+                                                      const attribute_id probe_attribute_id,
+                                                      const QueryContext::join_hash_table_id join_hash_table_index) {
+    std::vector<attribute_id> probe_attribute_ids;
+    probe_attribute_ids.push_back(probe_attribute_id);
```
We could refactor to `const std::vector<attribute_id> probe_attribute_ids(1, probe_attribute_id);`

saketj replied:
@zuyu Adding the const qualifier prevents the move semantics to be applied later in the function below, so dropping it.

zuyu commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,307 @@
+
+  std::unique_ptr<ExecutionHeuristics> execution_heuristics(new ExecutionHeuristics());
+  std::unique_ptr<QueryPlan> query_plan(new QueryPlan());
+  std::unique_ptr<serialization::QueryContext> query_context_proto(new serialization::QueryContext());
```
What do you think of moving these three variables in `ExecutionHeuristicsTest`, and then reset them in `SetUp`? Thus, we don't have to duplicate these code in each test fixture.

saketj replied:
Thanks. Moved them inside the Setup() now.

zuyu commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,307 @@
+
+  const attribute_id build_attribute_id_1 = 0;
+  const attribute_id build_attribute_id_2 = 0;
+  const attribute_id build_attribute_id_3 = 0;
```
I was wondering why we give three build attributes with the same id `0`?

saketj replied:
The build_attribute ids are inconsequential to the actual tests being performed. They are more of being used as placeholders.

zuyu commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,307 @@
+
+  // Test that the DAG was modified correctly or not.
+  // Probe operator 1 should have now build operator 1 and build operator 2 added as dependencies.
+  auto node_dependencies = query_plan->getQueryPlanDAG().getDependencies(probe_operator_index_1);
```
Mark `const`.

saketj replied:
Updated.

zuyu commented on storage/HashTable.hpp:
```diff
@@ -2164,6 +2247,27 @@ void HashTable<ValueT, resizable, serializable, force_key_copy, allow_duplicate_
+        bool bloom_miss = false;
+        for (std::size_t i = 0; i < probe_bloom_filters_.size() && !bloom_miss; ++i) {
+          const BloomFilter *bloom_filter = probe_bloom_filters_.at(i);
+          for (const attribute_id &attr_id : probe_attribute_ids_.at(i)) {
```
For above two `at(i)`s, we should use `[i]` for performance reasons.

saketj replied:
Thanks. Updated.

zuyu commented on storage/HashTableFactory.hpp:
```diff
@@ -305,10 +309,37 @@ class HashTableFactory {
+        }
+        hash_table->addProbeSideAttributeIds(std::move(probe_attribute_ids));
+      }
+    }
```
What do you think of moving all these late-initializations into `CreateResizable` and the hash table constructors? With this being said, we don't need additional helper methods in `HashTable.hpp`, and could mark these data members `const` wherever possible.

```
    bool has_build_side_bloom_filter = false;
    BloomFilter *build_bloom_filter = nullptr;
    if (proto.build_side_bloom_filters_size() > 0) {
      has_build_side_bloom_filter = true;
      build_bloom_filter = bloom_filters[proto.build_side_bloom_filter_id(0)].get();
    }

    const bool has_probe_side_bloom_filter = proto.probe_side_bloom_filters_size() > 0;
    std::vector<const BloomFilter*> probe_bloom_filters(proto.probe_side_bloom_filters_size());
    std::vector<std::vector<attribute_id>> probe_attribute_ids_list;
    for (int i = 0; i < proto.probe_side_bloom_filters_size(); ++i) {
      const auto probe_side_bloom_filter = proto.probe_side_bloom_filters(i);
      probe_bloom_filters[i] = bloom_filters[probe_side_bloom_filter.probe_side_bloom_filter_id()].get();

      std::vector<attribute_id> probe_attribute_ids(probe_side_bloom_filter.probe_side_attr_ids_size());
      for (int j = 0; j < probe_side_bloom_filter.probe_side_attr_ids_size(); ++j) {
        probe_attribute_ids[j] = probe_side_bloom_filter.probe_side_attr_ids(j);
      }
      probe_attribute_ids_list.push_back(std::move(probe_attribute_ids));
    }

    return CreateResizable(HashTableImplTypeFromProto(proto.hash_table_impl_type()),
                           key_types,
                           proto.estimated_num_entries(),
                           has_build_side_bloom_filter,
                           build_bloom_filter,
                           has_probe_side_bloom_filter,
                           std::move(probe_bloom_filters),
                           std::move(probe_attribute_ids_list),
                           storage_manager);
```

We also need to modify various hash table constructors, i.e., HashTable below:

```
// storage/HashTable.hpp
  HashTable(...,
            const bool has_build_side_bloom_filter,
            BloomFilter *build_bloom_filter,
            const bool has_probe_side_bloom_filter,
            std::vector<const BloomFilter*> &&probe_bloom_filters,
            std::vector<std::vector<attribute_id>> &&probe_attribute_ids)
      : ...
        has_build_side_bloom_filter_(has_build_side_bloom_filter),
        build_bloom_filter_(build_bloom_filter),
        has_probe_side_bloom_filter_(has_probe_side_bloom_filter),
        probe_bloom_filters_(std::move(probe_bloom_filters)),
        probe_attribute_ids_(std::move(probe_attribute_ids)) {}

...

  const bool has_build_side_bloom_filter_;
  BloomFilter *build_bloom_filter_;

  const bool has_probe_side_bloom_filter_;
  const std::vector<const BloomFilter*> probe_bloom_filters_;
  const std::vector<std::vector<attribute_id>> probe_attribute_ids_;
```

saketj replied:
@zuyu I also thought about pushing the initialization of these members all the way down to the constructor. However, I realized that such a change will lead to a large number of modifications to the constructor of all the specific hash table implementations, and in a way some code will be duplicated at each constructor of those hash table implementations. So, I thought this may be a more cleaner way to do it at one place, since this is common to all the instances of hash tables created by the factory.

zuyu replied:
No, we won't have duplicated implementations at all. Check out how `InsertDestination` handles the same situation.

What we need to do is for various hash table constructors, to pass multiple arguments along the way.

One benefit of this approach is that we could get rid of several set-up helper methods, and mark the data members `const`.

saketj replied:
@zuyu Had a discussion about this in Monday meeting today and we have agreed to pick this code refactoring as a part of a future PR, since it does not affect the current performance benchmarking required for the paper deadline. Marking this as a TODO for now. Thanks.

zuyu commented on utility/BloomFilter.hpp:
```diff
@@ -61,39 +91,168 @@ class BloomFilter {
+   * @param bloom_filter_proto The protobuf representation of a
+   *        bloom filter configuration.
+   **/
+  explicit BloomFilter(const serialization::BloomFilter bloom_filter_proto)
```
Please use the reference (`&`) here to avoid a deep copy. Thanks.

saketj replied:
Thanks. Updated.

zuyu commented on utility/BloomFilter.hpp:
```diff
@@ -61,39 +91,168 @@ class BloomFilter {
+    if (bloom_filter.bloom_filter_size() <= 0) {
+      return false;
+    }
+    return true;
```
We should refactor as following:

```
  static bool ProtoIsValid(const serialization::BloomFilter &proto) {
    return proto. IsInitialized();
  }
```

Note that we don't need to check whether these two values are negative, as we are using unsigned values for them.

saketj replied:
Refactored.

hbdeshmukh commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,301 @@
+
+  // Test that the DAG was modified correctly or not.
+  // Probe operator 1 should have now build operator 1 and build operator 2 added as dependencies.
+  auto const node_dependencies = query_plan_->getQueryPlanDAG().getDependencies(probe_operator_index_1);
```
Please rename to ``probe_node_dependencies`` or something similar for clarity.

hbdeshmukh commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,301 @@
+  // Test that the DAG was modified correctly or not.
+  // Probe operator 1 should have now build operator 1 and build operator 2 added as dependencies.
+  auto const node_dependencies = query_plan_->getQueryPlanDAG().getDependencies(probe_operator_index_1);
+  EXPECT_EQ(1u, node_dependencies.count(build_operator_index_2));
```
We should perform a similar check before modifying the DAG.

hbdeshmukh commented on query_optimizer/tests/ExecutionHeuristics_unittest.cpp:
```diff
@@ -0,0 +1,301 @@
+}
+
+TEST_F(ExecutionHeuristicsTest, HashJoinNotOptimizedTest) {
+  // This test case creates three hash joins, all of which are being probed on the different relation.
```
Please make the comment consistent with this test fixture.

zuyu commented on utility/CMakeLists.txt:
```diff
@@ -146,6 +146,10 @@ configure_file (
+QS_PROTOBUF_GENERATE_CPP(quickstep_utility_BloomFilter_proto_srcs
+                         quickstep_utility_BloomFilter_proto_hdrs
+                         BloomFilter.proto)
+
```
We don't need this extra empty line. Please remove it.

hbdeshmukh commented:
FYI, I cancelled the Travis build.

hbdeshmukh commented:
LGTM, merging.
